tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/join/ControlledWorkerBindJoinBase.java

```diff
@@ -0,0 +1,163 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Eclipse RDF4J contributors.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *******************************************************************************/
+package org.eclipse.rdf4j.federated.evaluation.join;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.rdf4j.common.iteration.CloseableIteration;
+import org.eclipse.rdf4j.federated.algebra.BoundJoinTupleExpr;
+import org.eclipse.rdf4j.federated.evaluation.FederationEvalStrategy;
+import org.eclipse.rdf4j.federated.evaluation.concurrent.ControlledWorkerScheduler;
+import org.eclipse.rdf4j.federated.evaluation.concurrent.ParallelExecutor;
+import org.eclipse.rdf4j.federated.evaluation.concurrent.ParallelTask;
+import org.eclipse.rdf4j.federated.structures.QueryInfo;
+import org.eclipse.rdf4j.query.BindingSet;
+import org.eclipse.rdf4j.query.QueryEvaluationException;
+import org.eclipse.rdf4j.query.algebra.TupleExpr;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Base class for executing joins as bind joins (i.e., the bindings of a block are injected in the SPARQL query as
+ * VALUES clause).
+ *
+ * The number of concurrent threads is controlled by a {@link ControlledWorkerScheduler} which works according to the
+ * FIFO principle and uses worker threads.
+ *
+ * This join cursor blocks until all scheduled tasks are finished, however the result iteration can be accessed from
+ * different threads to allow for pipelining.
+ *
+ * @author Andreas Schwarte
+ *
+ */
+public abstract class ControlledWorkerBindJoinBase extends JoinExecutorBase<BindingSet> {
+
+	private static final Logger log = LoggerFactory.getLogger(ControlledWorkerBindJoinBase.class);
+
+	protected final ControlledWorkerScheduler<BindingSet> scheduler;
+
+	protected final Phaser phaser = new Phaser(1);
```

Comment:
I've seen in other code that we `phaser.forceTermination()` in the `handleClose` method. Would that be wise here too?

Reply:
good observation, seems present in the code, see line 125 (not sure if I added in the force-push, but likely it was already present before)